rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+
+**TODO**: clarify what this last sentence exactly means.
+
+### Nickel and Nixpkgs
+
+As underlined in [Interaction with Nixpkgs](#interaction-with-nixpkgs), an FFI
+mechanism would require a bidirectional communication between Nix and Nickel.
+This is made complicated by the nature of both languages, which are lazy and
+functional, meaning that a Nix function called from Nickel may incur arbitrary
+calls back to Nickel (and vice-versa) to apply functions or to evaluate thunks.
+Not only does this seem complex, but also hard to be made efficient.
+
+Another possibility is to have everything evaluate in only one world, either Nix
+or Nickel. Practically, we could compile the Nix code to Nickel or vice-versa,
+and run everything on one side only.
+
+Compiling Nickel to Nix and run everything on the Nix side would have the
+advantage of leveraging all the Nix primitives and machinery freely. However,
+this solution simply nullifies a lot of the value proposition of Nickel. The
+added value of contracts is mainly due to its careful design as a primitive
+feature of the language, and while it could be simulated to some extent in Nix,
+the error message and performances just couldn't be in par. Similarly, the merge
+system being native is in part motivated by the perspective of better
+performance and error message. This simply becomes moot if we compile those
+features away back to plain Nix code.
+
+On the other hand, compiling Nix to Nickel looks more appealing. Nickel is close
+to being a superset of Nix (the only non-trivial missing features being the
+`with` construct and string contexts, but the latter doesn't even really change
+the semantics). This would preserve the Nickel advances, and we could leverage
+directly any non-builtin Nix function, including all the Nixpkgs library
+functions.
+
+The potential drawbacks of this approach:
+
+- Performance: Nickel is not expected to be in par with Nix at this point, so
+  the performance of Nixpkgs compiled to Nickel might be prohibitively bad at
+  first. On a positive note, this would provide a great benchmark suite and
+  motivation to guide future performance improvements.
+- Some builtin may incur effects, which are not trivial to combine with the
+  execution model (revertible thunks). That being said, those effects are in
+  practice idempotent and commutative (creating paths in the store, creating
+  derivation, etc.), which should be fine.
+
+Compilation could be done on the fly, as the compilation process would be rather
+straightforward, or have a Nixpkgs snapshot pre-compiled to Nickel, or a mix of
+both.
+
+This solution requires to reimplement Nix builtins in Nickel (the nickel-nix
+compatibility layer) . This is an interesting milestone in itself, because even
+without a Nix-to-Nickel compiler, the compatibility layer would already make
+writing derivations in pure Nickel possible.
+
+- TODO: what about dynamic imports?
+- TODO: what about using a Nickel pkgs from Nix?
```

Comment:
Unless (until?) we can get to a point where the Nix evaluator is just Nickel with a Nix frontend (in which case that problem won’t exist anymore because Nix could just call Nickel code transparently), I think that could be a limiting factor for the adoption of Nickel − at least for open-source projects, but these are the one with the most visibility. So we’ll want to have a decent story around this too

Reply:
As mentioned during the meeting, it seems going from a data representation to a function representation is easier, so we can hope for a simple generic wrappers around a Nickel package. But that needs to be investigated further before proceeding.

Reply:
Are we back to considering a functional representation for Nickel pkgs?

Reply:
I think what is being discussed here is having a wrapper of some sort to use a Nickel package from Nix code, but just as a legacy compatibility mechanism. Using Nickel packages from another Nickel package would use the data representation and leverage all the benefits. However, if for some reason you need to use a package defined in Nickel from a Nixpkgs-like Nix codebase, you could, thanks to (very schematically) `to_nixpkg : Package -> NixPkgLegacyFunction`.